allennlp/commands/find_learning_rate.py

```diff
@@ -193,6 +193,14 @@ def find_learning_rate_model(
     prepare_environment(params)
 
     cuda_device = params.params.get("trainer").get("cuda_device", -1)
+    devices = parse_cuda_device(cuda_device)
+
+    # HACK: The trainer can not be constructed with multiple gpus.
```

Comment:
@brendan-ai2 - this PR makes the `Trainer` take strictly a single GPU. This presents a small problem, which is that we still want configs which have a `"cuda_device": [1,2,3]` line to work with other commands which are not `train` (like in this find_learning_rate command). 

I put this hack in here to demonstrate the problem (we have to make sure to manually parse the cuda devices every time we try to build the trainer, which is kind of gross), but I think a good way to resolve this is to have a separate `"distributed_gpus"` argument, either in the trainer or the top level of the config, which we can use in `train` to determine and launch the workers. This way, `Trainer.from_params` would be guaranteed to work in all cases without manually parsing out the cuda device.

Also, maybe it makes sense to move the `distributed`, `master_address` and `master_port` config arguments to the top level of the config - they are not actually arguments to the `Trainer`, they just affect how it is constructed within the `train` command. Does that make sense? Do you have any other ideas?

Reply:
Actually I think moving the distributed config up to the top level just solves all the problems, I think i'll do that.

Reply:
Yeah, that makes total sense. One suggestion though, I suspect that we may need further distributed options soon. (E.g., various multi-node settings) What about having config like:

```
"trainer": {"num_epochs": 2, "optimizer": "adam"},
"distributed": {"cuda_devices": [0, 1, 2]},
```

This gives us an obvious place to add them.